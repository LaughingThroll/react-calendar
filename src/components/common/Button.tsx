import React, { ReactElement } from 'react'
import classnames from 'classnames'

interface IButton extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  secondary?: boolean
  classNames?: string[]
  type?: 'button' | 'submit' | 'reset'
  iconPlus?: boolean
  onClick?: (...args: any) => any
  children?: string | ReactElement
  disabled?: boolean
}

const Button: React.FC<IButton> = ({
  type = 'button',
  children,
  secondary,
  disabled,
  classNames = [],
  iconPlus,
  ...rest
}) => {
  return (
    <button
      type={type}
      className={`${classnames({
        button: true,
        'button--icon-plus': iconPlus,
        'button--secondary': secondary,
        'button--primary': !secondary,
        'button--disabled': disabled,
      })} ${classNames.join(' ')}`}
      {...rest}
    >
      {children}
    </button>
  )
}

export default React.memo(Button)
