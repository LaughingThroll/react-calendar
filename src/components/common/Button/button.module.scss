@import 'styles/utils/variables';
@import 'styles/utils/mixin';

.button {
  @extend %font-bold;

  padding: 10px;
  min-width: 144px;
  border-radius: 7px;
  display: flex;
  align-items: center;
  justify-content: center;
  transition: background-color $fast linear, color $fast linear;

  &_primary {
    background-color: $button--primary;
    color: $white;
  }

  &_secondary {
    border: 1px solid $button--secondary-border;
    color: $scorpion;
  }

  &_disabled {
    pointer-events: none;
    background-color: $button--disabled;
  }

  &_icon_plus {
    &::before {
      content: '';
      background-image: url('../../../assets/images/icons/plus.svg');
      background-position: center center;
      background-size: cover;
      background-repeat: no-repeat;
      width: 13px;
      height: 13px;
      margin-right: 10px;
    }
  }
}
