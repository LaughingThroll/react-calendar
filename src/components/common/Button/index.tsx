import React from 'react'
import classnames from 'classnames'
import styles from './button.module.scss'
import { TChildren } from '../../../types/utilsTypes'

export interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  type?: 'button' | 'submit' | 'reset'
  classNames?: string[]
  secondary?: boolean
  iconPlus?: boolean
  disabled?: boolean
  onClick?: (...args: any) => any
  children?: TChildren
}

const Button: React.FC<ButtonProps> = ({
  type = 'button',
  children,
  secondary,
  disabled,
  classNames = [],
  iconPlus,
  ...rest
}) => {
  const { button, button_icon_plus, button_secondary, button_primary, button_disabled } = styles
  return (
    <button
      type={type}
      className={`${classnames({
        [button]: true,
        [button_icon_plus]: iconPlus,
        [button_secondary]: secondary,
        [button_primary]: !secondary,
        [button_disabled]: disabled,
      })} ${classNames.join(' ')}`}
      {...rest}
    >
      {children}
    </button>
  )
}

export default React.memo(Button)
