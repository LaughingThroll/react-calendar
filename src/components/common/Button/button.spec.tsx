import React from 'react'
import { shallow } from 'enzyme'
import Button, { ButtonProps } from './'

const setUp = (props: ButtonProps) => shallow(<Button {...props} />)

describe('Button component initial', () => {
  it('render button', () => {
    const tree = setUp({})
    expect(tree.find('button')).toHaveLength(1)
  })
  it('render button without props', () => {
    const tree = setUp({})

    expect(tree).toMatchSnapshot()
  })

  it('button with props default', () => {
    const tree = setUp({})

    expect(tree.find('button').prop('className')).toEqual('button button_primary ')
  })
})

describe('Button component testing props "type"', () => {
  it('default props "type"', () => {
    const tree = setUp({})

    expect(tree.find('button').prop('type')).toEqual('button')
  })

  it('props "type" submit', () => {
    const tree = setUp({ type: 'submit' })

    expect(tree.find('button').prop('type')).toEqual('submit')
  })

  it('props "type" reset', () => {
    const tree = setUp({ type: 'reset' })

    expect(tree.find('button').prop('type')).toEqual('reset')
  })
})

describe('Button component testing with props', () => {
  it('props "classNames"', () => {
    const tree = setUp({ classNames: ['hello', 'world'] })

    expect(tree.find('button').prop('className')).toEqual('button button_primary hello world')
  })

  it('props "secondary"', () => {
    const tree = setUp({ secondary: true })

    expect(tree.find('button').prop('className')).toEqual('button button_secondary ')
  })

  it('props "iconPlus"', () => {
    const tree = setUp({ iconPlus: true })

    expect(tree.find('button').prop('className')).toEqual('button button_icon_plus button_primary ')
  })

  it('props "disabled"', () => {
    const tree = setUp({ disabled: true })

    expect(tree.find('button').prop('className')).toEqual('button button_primary button_disabled ')
  })

  it('props "onClick" without args', () => {
    // Preparing
    const props = { onClick: jest.fn() }
    const tree = setUp(props)
    const $button = tree.find('button')

    // ACT
    $button.simulate('click')

    // Expectation
    expect(props.onClick).toHaveBeenCalled()
  })

  it('props "onClick" with args', () => {
    // Preparing
    const props = { onClick: jest.fn((x) => x + 1) }
    const tree = setUp(props)
    const $button = tree.find('button')

    // ACT
    $button.simulate('click')

    // Expectation
    expect(props.onClick).toHaveBeenCalled()
    expect(props.onClick(1)).toEqual(2)
  })
})
