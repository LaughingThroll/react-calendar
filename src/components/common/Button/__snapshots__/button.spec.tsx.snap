// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component initial render button without props 1`] = `
<button
  className="button button_primary "
  type="button"
/>
`;
